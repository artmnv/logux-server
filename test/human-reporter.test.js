'use strict'

const processReporter = require('../reporters/bunyan/process')

const ServerConnection = require('logux-sync').ServerConnection
const createServer = require('http').createServer
const SyncError = require('logux-sync').SyncError
const path = require('path')
const bunyan = require('bunyan')

const ServerClient = require('../server-client')
const BaseServer = require('../base-server')

const BunyanFormatStream = require('../reporters/human/format')
const MemoryStream = require('memory-stream')

const DATE = /\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d/g

function bunyanLog (logger, payload) {
  const details = payload.details || { }
  logger[payload.level](details, payload.msg)
}

function reportersOut (type, app) {
  const payload = processReporter.apply({ }, arguments)
  const memStream = new MemoryStream()
  const formatOut = new BunyanFormatStream(app, memStream)
  const bunyanLogger = bunyan.createLogger({
    name: 'logux-server-test',
    stream: formatOut
  })

  bunyanLog(bunyanLogger, payload)

  return new Promise(resolve => {
    setTimeout(() => {
      const result = memStream
        .toString()
        .replace(/\r\v/g, '\n')
        .replace(DATE, '1970-01-01 00:00:00')
        .replace(/PID:.+\n/g, 'PID:          21384\n')
      resolve(result)
    }, 50)
  })
}

const log = bunyan.createLogger({
  name: 'logux-server-test',
  streams: []
})

const app = new BaseServer({
  env: 'development',
  pid: 21384,
  subprotocol: '2.5.0',
  supports: '2.x || 1.x',
  host: '127.0.0.1',
  port: 1337,
  reporter: 'bunyan',
  bunyanLogger: log
})
app.nodeId = 'server:H1f8LAyzl'

const ws = {
  _socket: {
    remoteAddress: '127.0.0.1'
  },
  on: () => { }
}

const authed = new ServerClient(app, new ServerConnection(ws), 1)
authed.sync.remoteSubprotocol = '1.0.0'
authed.sync.remoteProtocol = [0, 0]
authed.id = '100'
authed.user = { }
authed.nodeId = '100:H10Nf5stl'

const noUserId = new ServerClient(app, new ServerConnection(ws), 1)
noUserId.sync.remoteSubprotocol = '1.0.0'
noUserId.sync.remoteProtocol = [0, 0]
noUserId.user = { }
noUserId.nodeId = 'H10Nf5stl'

const unauthed = new ServerClient(app, new ServerConnection(ws), 1)

const ownError = new SyncError(authed.sync, 'timeout', 5000, false)
const clientError = new SyncError(authed.sync, 'timeout', 5000, true)

const action = {
  type: 'CHANGE_USER',
  id: 100,
  data: { name: 'John' }
}
const meta = {
  id: [1487805099387, authed.nodeId, 0],
  time: 1487805099387,
  reasons: ['lastValue', 'debug'],
  user: authed.nodeId,
  server: 'server:H1f8LAyzl'
}

it('reports listen', () => {
  return reportersOut('listen', app).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports bad log info', () => {
  const memStream = new MemoryStream()
  const formatOut = new BunyanFormatStream(
    { outputMode: 'logux' }, memStream, app
  )

  formatOut.write('Simple text payload')

  return new Promise(resolve => {
    setTimeout(() => {
      const result = memStream
        .toString()
        .replace(/\r\v/g, '\n')
        .replace(DATE, '1970-01-01 00:00:00')
      resolve(result)
    }, 50)
  }).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports production', () => {
  const wss = new BaseServer({
    env: 'production',
    pid: 21384,
    subprotocol: '1.0.0',
    supports: '1.x',
    cert: 'A',
    key: 'B',
    host: '0.0.0.0',
    port: 1337
  })
  wss.nodeId = 'server:H1f8LAyzl'

  return reportersOut('listen', wss).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports http', () => {
  const http = new BaseServer({
    env: 'development',
    pid: 21384,
    subprotocol: '1.0.0',
    supports: '1.x',
    server: createServer()
  })
  http.nodeId = 'server:H1f8LAyzl'

  return reportersOut('listen', http).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports connect', () => {
  return reportersOut('connect', app, authed).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports authenticated', () => {
  return reportersOut('authenticated', app, authed).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports authenticated without user ID', () => {
  return reportersOut('authenticated', app, noUserId).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports bad authenticated', () => {
  return reportersOut('unauthenticated', app, authed).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports action', () => {
  return reportersOut('add', app, action, meta).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports clean', () => {
  return reportersOut('clean', app, action, meta).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports denied', () => {
  return reportersOut('denied', app, action, meta).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports processed', () => {
  return reportersOut('processed', app, action, meta, 500).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports unknownType', () => {
  return reportersOut('unknownType', app, action, meta).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports unknownType from server', () => {
  const serverMeta = { id: [1, 'server:hfeb5', 0] }
  return reportersOut('unknownType', app, action, serverMeta).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports disconnect', () => {
  return reportersOut('disconnect', app, authed).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports disconnect from unauthenticated user', () => {
  return reportersOut('disconnect', app, unauthed).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports error', () => {
  const file = __filename
  const jest = path.join(__dirname, '../node_modules/jest/index.js')
  const error = new Error('Some mistake')
  const errorStack = [
    `${ error.name }: ${ error.message }`,
    `    at Object.<anonymous> (${ file }:28:13)`,
    `    at Module._compile (module.js:573:32)`,
    `    at at runTest (${ jest }:50:10)`,
    `    at process._tickCallback (internal/process/next_tick.js:103:7)`
  ]
  error.stack = errorStack.join('\n')

  const out = reportersOut('runtimeError', app, error, action, meta)
  return out.then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports client error', () => {
  const out = reportersOut('clientError', app, authed, clientError)
  return out.then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports synchronization error', () => {
  const out = reportersOut('syncError', app, authed, ownError)
  return out.then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports error from unautheficated user', () => {
  const out = reportersOut('syncError', app, unauthed, clientError)
  return out.then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports zombie', () => {
  return reportersOut('zombie', app, authed).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('reports destroy', () => {
  return reportersOut('destroy', app).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('handles EACCES error', () => {
  return reportersOut('error', app, { code: 'EACCES' }, app).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('handles error in production', () => {
  const http = new BaseServer({
    env: 'production',
    pid: 21384,
    subprotocol: '2.5.0',
    supports: '2.x || 1.x',
    host: '127.0.0.1',
    port: 1000
  })
  http.nodeId = 'server:H1f8LAyzl'

  return reportersOut('error', app, { code: 'EACCES', port: 1000 }, http)
    .then(data => {
      expect(data).toMatchSnapshot()
    })
})

it('handles EADDRINUSE error', () => {
  return reportersOut(
    'error', app, {
      code: 'EADDRINUSE',
      port: 1337
    }, app).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('handles Logux initialization error: unknown option', () => {
  return reportersOut('error', app, {
    code: 'LOGUX_UNKNOWN_OPTION',
    option: 'test'
  }, app).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('handles Logux initialization error: subprotocol', () => {
  return reportersOut('error', app, {
    code: 'LOGUX_WRONG_OPTIONS',
    message: 'Missed client subprotocol requirements'
  }, app).then(data => {
    expect(data).toMatchSnapshot()
  })
})

it('throws on undefined error', () => {
  const e = {
    code: 'EAGAIN',
    message: 'resource temporarily unavailable'
  }
  function errorHelperThrow () {
    reportersOut('error', app, e)
  }
  expect(errorHelperThrow).toThrowError(/resource temporarily unavailable/)
})
