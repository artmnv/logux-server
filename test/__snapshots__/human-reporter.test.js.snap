// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles EACCES error 1`] = `
"[1m[31m[40m[7m ERROR [27m[49m[39m[22m [1m[31mYou are not allowed to run server on this port[39m[22m [2mat 1970-01-01 00:00:00[22m
        Try to change user (e.g. root) or use port >= 1024

"
`;

exports[`handles EADDRINUSE error 1`] = `
"[1m[31m[40m[7m ERROR [27m[49m[39m[22m [1m[31mPort :1337 already in use[39m[22m [2mat 1970-01-01 00:00:00[22m
        Another Logux server or other app already running on this port
        Maybe you didnâ€™t not stop server from other project
        or previous version of this server was not killed

"
`;

exports[`handles Logux initialization error: subprotocol 1`] = `
"[1m[31m[40m[7m ERROR [27m[49m[39m[22m [1m[31mMissed client subprotocol requirements[39m[22m [2mat 1970-01-01 00:00:00[22m
        Check server constructor and Logux Server documentation

"
`;

exports[`handles Logux initialization error: unknown option 1`] = `
"[1m[31m[40m[7m ERROR [27m[49m[39m[22m [1m[31mUnknown option \`test\` in server constructor[39m[22m [2mat 1970-01-01 00:00:00[22m
        Maybe there is a mistake in option name or this version
        of Logux Server doesnâ€™t support this option

"
`;

exports[`handles error in production 1`] = `
"[1m[31m[40m[7m ERROR [27m[49m[39m[22m [1m[31mYou are not allowed to run server on this port[39m[22m [2mat 1970-01-01 00:00:00[22m
        Try to change user (e.g. root) or use port >= 1024

"
`;

exports[`reports action 1`] = `
"[1m[32m[40m[7m INFO [27m[49m[39m[22m  [1m[32mAction was added[39m[22m [2mat 1970-01-01 00:00:00[22m
        Action: 
          type: [1mCHANGE_USER[22m
          id:   [1m100[22m
          data: 
            name: [1mJohn[22m
        Meta:   
          id:      [[1m1487805099387[22m, [1m100:H10Nf5stl[22m, [1m0[22m]
          time:    [1m1487805099387[22m
          reasons: [[1mlastValue[22m, [1mdebug[22m]
          user:    [1m100:H10Nf5stl[22m
          server:  [1mserver:H1f8LAyzl[22m

"
`;

exports[`reports authenticated 1`] = `
"[1m[32m[40m[7m INFO [27m[49m[39m[22m  [1m[32mUser was authenticated[39m[22m [2mat 1970-01-01 00:00:00[22m
        Node ID:     [1m100[22m:H10Nf5stl
        Subprotocol: [1m1.0.0[22m
        Client ID:   [1m1[22m

"
`;

exports[`reports authenticated without user ID 1`] = `
"[1m[32m[40m[7m INFO [27m[49m[39m[22m  [1m[32mUser was authenticated[39m[22m [2mat 1970-01-01 00:00:00[22m
        Node ID:     H10Nf5stl
        Subprotocol: [1m1.0.0[22m
        Client ID:   [1m1[22m

"
`;

exports[`reports bad authenticated 1`] = `
"[1m[33m[40m[7m WARN [27m[49m[39m[22m  [1m[33mBad authentication[39m[22m [2mat 1970-01-01 00:00:00[22m
        Node ID:     [1m100[22m:H10Nf5stl
        Subprotocol: [1m1.0.0[22m
        Client ID:   [1m1[22m

"
`;

exports[`reports bad log info 1`] = `"Simple text payload"`;

exports[`reports clean 1`] = `
"[1m[32m[40m[7m INFO [27m[49m[39m[22m  [1m[32mAction was cleaned[39m[22m [2mat 1970-01-01 00:00:00[22m
        Action ID: [[1m1487805099387[22m, [1m100:H10Nf5stl[22m, [1m0[22m]

"
`;

exports[`reports client error 1`] = `
"[1m[33m[40m[7m WARN [27m[49m[39m[22m  [1m[33mClient error: A timeout was reached (5000ms)[39m[22m [2mat 1970-01-01 00:00:00[22m
        Node ID: [1m100[22m:H10Nf5stl

"
`;

exports[`reports connect 1`] = `
"[1m[32m[40m[7m INFO [27m[49m[39m[22m  [1m[32mClient was connected[39m[22m [2mat 1970-01-01 00:00:00[22m
        Client ID:  [1m1[22m
        IP address: [1m127.0.0.1[22m

"
`;

exports[`reports denied 1`] = `
"[1m[33m[40m[7m WARN [27m[49m[39m[22m  [1m[33mAction was denied[39m[22m [2mat 1970-01-01 00:00:00[22m
        Action ID: [[1m1487805099387[22m, [1m100:H10Nf5stl[22m, [1m0[22m]

"
`;

exports[`reports destroy 1`] = `
"[1m[32m[40m[7m INFO [27m[49m[39m[22m  [1m[32mShutting down Logux server[39m[22m [2mat 1970-01-01 00:00:00[22m

"
`;

exports[`reports disconnect 1`] = `
"[1m[32m[40m[7m INFO [27m[49m[39m[22m  [1m[32mClient was disconnected[39m[22m [2mat 1970-01-01 00:00:00[22m
        Node ID: [1m100[22m:H10Nf5stl

"
`;

exports[`reports disconnect from unauthenticated user 1`] = `
"[1m[32m[40m[7m INFO [27m[49m[39m[22m  [1m[32mClient was disconnected[39m[22m [2mat 1970-01-01 00:00:00[22m
        Client ID: [1m1[22m

"
`;

exports[`reports error 1`] = `
"[1m[31m[40m[7m ERROR [27m[49m[39m[22m [1m[31mSome mistake[39m[22m [2mat 1970-01-01 00:00:00[22m
[33m        at Object.<anonymous> (test/human-reporter.test.js:28:13)[39m
[31m        at Module._compile (module.js:573:32)[39m
[31m        at at runTest (node_modules/jest/index.js:50:10)[39m
[31m        at process._tickCallback (internal/process/next_tick.js:103:7)[39m
        Action ID: [[1m1487805099387[22m, [1m100:H10Nf5stl[22m, [1m0[22m]

"
`;

exports[`reports error from unautheficated user 1`] = `
"[1m[31m[40m[7m ERROR [27m[49m[39m[22m [1m[31mSyncError from client: A timeout was reached (5000ms)[39m[22m [2mat 1970-01-01 00:00:00[22m
        Client ID: [1m1[22m

"
`;

exports[`reports http 1`] = `
"[1m[32m[40m[7m INFO [27m[49m[39m[22m  [1m[32mLogux server is listening[39m[22m [2mat 1970-01-01 00:00:00[22m
        PID:          21384
        Logux server: [1m0.1.1[22m
        Node ID:      [1mserver[22m:H1f8LAyzl
        Environment:  [1mdevelopment[22m
        Subprotocol:  [1m1.0.0[22m
        Supports:     [1m1.x[22m
        Listen:       [1mCustom HTTP server[22m
        [90mServer was started in non-secure development mode[39m
        [90mPress Ctrl-C to shutdown server[39m

"
`;

exports[`reports listen 1`] = `
"[1m[32m[40m[7m INFO [27m[49m[39m[22m  [1m[32mLogux server is listening[39m[22m [2mat 1970-01-01 00:00:00[22m
        PID:          21384
        Logux server: [1m0.1.1[22m
        Node ID:      [1mserver[22m:H1f8LAyzl
        Environment:  [1mdevelopment[22m
        Subprotocol:  [1m2.5.0[22m
        Supports:     [1m2.x || 1.x[22m
        Listen:       [1mws://127.0.0.1:1337[22m
        [90mServer was started in non-secure development mode[39m
        [90mPress Ctrl-C to shutdown server[39m

"
`;

exports[`reports processed 1`] = `
"[1m[32m[40m[7m INFO [27m[49m[39m[22m  [1m[32mAction was processed[39m[22m [2mat 1970-01-01 00:00:00[22m
        Action ID: [[1m1487805099387[22m, [1m100:H10Nf5stl[22m, [1m0[22m]
        Duration:  [1m500[22m

"
`;

exports[`reports production 1`] = `
" INFO   Logux server is listening at 1970-01-01 00:00:00
        PID:          21384
        Logux server: 0.1.1
        Node ID:      server:H1f8LAyzl
        Environment:  production
        Subprotocol:  1.0.0
        Supports:     1.x
        Listen:       wss://0.0.0.0:1337

"
`;

exports[`reports synchronization error 1`] = `
"[1m[31m[40m[7m ERROR [27m[49m[39m[22m [1m[31mSyncError: A timeout was reached (5000ms)[39m[22m [2mat 1970-01-01 00:00:00[22m
        Node ID: [1m100[22m:H10Nf5stl

"
`;

exports[`reports unknownType 1`] = `
"[1m[33m[40m[7m WARN [27m[49m[39m[22m  [1m[33mAction with unknown type CHANGE_USER[39m[22m [2mat 1970-01-01 00:00:00[22m
        Action ID: [[1m1487805099387[22m, [1m100:H10Nf5stl[22m, [1m0[22m]

"
`;

exports[`reports unknownType from server 1`] = `
"[1m[31m[40m[7m ERROR [27m[49m[39m[22m [1m[31mAction with unknown type CHANGE_USER[39m[22m [2mat 1970-01-01 00:00:00[22m
        Action ID: [[1m1[22m, [1mserver:hfeb5[22m, [1m0[22m]

"
`;

exports[`reports zombie 1`] = `
"[1m[33m[40m[7m WARN [27m[49m[39m[22m  [1m[33mZombie client was disconnected[39m[22m [2mat 1970-01-01 00:00:00[22m
        Node ID: [1m100[22m:H10Nf5stl

"
`;
